mat/qr.go

```diff
@@ -98,23 +98,9 @@ func (qr *QR) factorize(a Matrix, norm lapack.MatrixNorm) {
 	lapack64.Geqrf(qr.qr.mat, qr.tau, work, len(work))
 	putFloat64s(work)
 	qr.updateCond(norm)
-	qr.updateQ()
```

Comment:
I'm afraid this doesn't work because `q` is currently needed in the `At` method. The test passes just because `QTo` is called in the test before `EqualApprox` (which calls `At`).

We could reconstruct the necessary row of Q in each call to `At` which is terrible but hopefully nobody uses `QR` as `Matrix` except in the call to `Solve`?

Reply:
The other alternative is to lazily calculate Q for `At` the same way it is for `ToQ`, with a warning that it may cause OoM. I think your approach is probably better. A warning in the docs that it will be extremely inefficient should be enough.

Reply:
We could lazily compute `q` in the `At()` method if `q` is nil or empty, in the same way it is done in the `Qto`.

But I guess I'm also a bit confused as to why we need `q` at all in `At()` in the first place. When we factorize a matrix `a`, `qr` stores a copy of that matrix. Why not return directly `qr.At()`?

Isn't https://github.com/gonum/gonum/blame/2ad11cabb395b96efc5b67fa1b64480762d9e703/mat/qr.go#L46 also not faulty?
I'm expected we should return Q*R at element(i,j), but instead it seems we're returning Q*A at element (i,j), with A = Q*R, the matrix we factorize.

Reply:
> The other alternative is to lazily calculate Q for At the same way it is for ToQ.

Agreed. If one element is needed, then most likely all are.

> qr stores a copy of that matrix. Why not return directly qr.At()?

That copy is overwritten by `lapack64.Geqrf` which efficiently stores both Q and R in the same matrix.

Reply:
Another alternative is to indeed store `a` in `Factorize` and compute Q in `QTo` without storing it.

Reply:
I think we would need to copy `a` in this situation.

Reply:
In the end I think it's best to lazily compute Q in At. QR.At should never be called in practice.

Reply:
Any chance this is called inadvertently by using QR as a matrix input to another function (since it now fulfills the `Matrix` interface), e.g. `Trace()`.

Reply:
I'm thinking that @tvkn's suggestion of conditionally computing element of QR for `At` via the relevant row of Q if Q is not already calculated, otherwise falling through to the code that is there now. 
```
diff --git a/mat/qr.go b/mat/qr.go
index d4787bc3..7caa2f20 100644
--- a/mat/qr.go
+++ b/mat/qr.go
@@ -41,6 +41,11 @@ func (qr *QR) At(i, j int) float64 {
 		panic(ErrColAccess)
 	}
 
+	if qr.q == nil || qr.q.IsEmpty() {
+		// Calculate Q's row i;
+		// Return Q[i,]×R[,j].
+	}
+
 	var val float64
 	for k := 0; k <= j; k++ {
 		val += qr.q.at(i, k) * qr.qr.at(k, j)
```
This is safer than just lazily calculating Q which has the risk of someone thinking the code is fine until a tall matrix is used down the track. At worst with this approach there will be a perf hit, but not a crash.

I don't have time to do this, so if someone wants to pick this up. I'm happy for that to happen.